(wip) tr1: do matrix projections on the gpu

rr- opened the pull request:
#### Checklist

- [x] I have read the [coding conventions](https://github.com/LostArtefacts/TRX/blob/master/CONTRIBUTING.md#coding-conventions)
- [x] I have added a changelog entry about what my pull request accomplishes, or it is an internal change
- [x] I have added a readme entry about my new feature or OG bug fix, or it is a different change

#### Description

This is a heavy WIP where we simplify TR1 rendering to do the matrix transforms on the GPU. The eventual goal is less code and more performant pipeline.
Out of necessity, this moves wibble transform to the GPU as well.

Known problems:
- [x] Seams along room borders. This is likely either due to the precision issues with the scissor test. The scissor test, introduced in this PR, discards every fragment outside a specific rectangle in screen coordinates, and we apply it here on the portals to deal with overlapping rooms.
- [ ] Sprites
    - [x] Geometry
    - [x] Camera projection
    - [x] Texturing
    - [x] Updating lighting
    - [x] Animating textures
    - [ ] Dynamic sprites (bubbles, flames) 
- [x] The inventory ring has Z Buffer problems, as usual.
- [x] Pickup items UI wrong projection
- [x] Debug mode
- [x] Fog is not working
- [x] Seams while Lara is underwater are back
- [x] Broken passport in TRUB (`/play 1`, `/title`, open passport)
- [x] Missing texture on the back of the passport in TRUB (`/play 1`, `/title`, open passport)
- [x] Framerate drop when Lara is standing in this spot in Return to Egypt.
- [ ] Windowed size problems
- [ ] Lara's shadow clips above water
- [ ] Lighting Z problems
- [ ] ~~The Save Game/Load Game headers in the save/load dialogs are missing~~ (present in develop)
- [ ] The room sprites are flickering, but only if bilinear filter is enabled
- [ ] Parts of Lara doesn't show reflection when she stands on Midas' hand, if the reflection option is enabled
- [ ] There's a subtle difference with the lighting

For now I'm looking for an early feedback about more broken things. I tried my best to maintain 100% compatibility with the wibble animation – both the shadings, and the geometry offsetting. LMK if it's accurate enough.

rr- committed:
(wip) projection on gpu

aredfan commented:
At a quick glance, the only issue I found is the 3d pickup animation is partially visible onscreen before sliding in, and doesn't fully slide out of view before disappearing.

**Edit:** I spoke too soon, take a look at debug. 😄 

![20250313_213149_Caves](https://github.com/user-attachments/assets/3985ac1c-afc6-444d-9e6a-3579714318d0)

lahm86 commented:
Looking good so far. The only additional thing I've noticed is missing fog. I can continue testing tomorrow.
![fog](https://github.com/user-attachments/assets/e59d33d1-d709-44d2-ba7e-e9d0e7c13697)

walkawayy commented:
I don't think FX bubbles are being clipped properly underwater. Try this spot in Vilcabamba. The bubbles only appear above the far wall. They also behave weirdly in the rest of the tunnel with some odd clipping in the geometry.

![20250313_183119_City_of_Vilcabamba](https://github.com/user-attachments/assets/e2a70acf-02e2-4f07-9656-3c2885f2ea0f)

rr- commented:
> I don't think FX bubbles are being clipped properly underwater. Try this spot in Vilcabamba. The bubbles only appear above the far wall. They also behave weirdly in the rest of the tunnel with some odd clipping in the geometry.
> 
> ![20250313_183119_City_of_Vilcabamba](https://private-user-images.githubusercontent.com/81546780/422561223-e2a70acf-02e2-4f07-9656-3c2885f2ea0f.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDE5MDU1MzAsIm5iZiI6MTc0MTkwNTIzMCwicGF0aCI6Ii84MTU0Njc4MC80MjI1NjEyMjMtZTJhNzBhY2YtMDJlMi00ZjA3LTk2NTYtM2MyODg1ZjJlYTBmLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzEzVDIyMzM1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTgzNTllMzdhYzBlYjQwNWJmNjZlNTBhOTQyOWY0OThkMGI5NDZhYmIzN2E3ZWIyZWEwMTI4NWI0MzNjNTBmMDAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.DQsrzjTsEq75DCuNHCdgts1-gDtn6azhIs9lcrZKiu4)

That's right – the bubbles are also sprites, and sprites aren't working just yet.

I've addressed all of the issues reported thus far apart from sprites.

aredfan commented:
- [ ] ~~I forget to mention this before regarding debug mode - I noticed a red border around the screen, and red boxes connected to the portals.~~

![20250314_172206_Caves](https://github.com/user-attachments/assets/3fefcfe1-075e-4115-96c2-efbbd94d4598)

- [x] It looks like the seams while Lara is underwater are back.


#1661, for reference. 

![Screenshot](https://github.com/user-attachments/assets/a8c0d2f6-59df-4ca0-a618-8b4fdaf5ca0b)

- [x] There appears to be an issue with the passport in TRUB.

1. launch trub
2. /play 1 (can be any level)
3. /title
4. open the passport

![20250314_181355_Level_0](https://github.com/user-attachments/assets/f914f5b5-5a1d-4893-9cb8-2d6eefdb9221)

There's a missing texture on the back of the passport when following the same steps.
![20250314_182218_Level_0](https://github.com/user-attachments/assets/2b79d478-24b6-4e7b-8be9-e5ea60638ffa)

- [x] A potential issue, there is a large framerate drop when Lara is standing in this spot in Return to Egypt.

![20250314_190318_Return_to_Egypt](https://github.com/user-attachments/assets/766890b7-e6bf-4ae0-a6ab-fea5463fb7a6)

rr- commented:
> * [ ]  I forget to mention this before regarding debug mode - I noticed a red border around the screen, and red boxes connected to the portals.
> 
> ![20250314_172206_Caves](https://private-user-images.githubusercontent.com/94776985/422897669-3fefcfe1-075e-4115-96c2-efbbd94d4598.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDE5ODI0ODgsIm5iZiI6MTc0MTk4MjE4OCwicGF0aCI6Ii85NDc3Njk4NS80MjI4OTc2NjktM2ZlZmNmZTEtMDc1ZS00MTE1LTk2YzItZWZiYmQ5NGQ0NTk4LmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzE0VDE5NTYyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTMyMGMyYzQyNDQ4Y2Y4MDkzNWNiZjM1NzYwNjU1OWVkMTAyYjc3ZjhmMTVmZGVkMzIzZDY4Yzc5YmNkNTI2ZjgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.iXLnGZwGgFjoX9NgvHeGAqTSJjXkxdusrXN_zAXO8tw)

The red frames are actually a feature and are used to debug portals.

aredfan commented:
- [ ] If the render mode is set to window size, there is an issue with the game's scaling which isn't present in develop.

1. start up the game in windowed mode, smaller than your resolution
2. play level 0
3. open the inventory

![Screenshot](https://github.com/user-attachments/assets/ab979ac8-957e-477f-aa79-873805d41254)

- [ ] A small issue that might be relevant. When Lara is flying above water, it appears as if her shadow is clipping at the water's surface.

rr- committed:
misc: fix small memory leaks

aredfan commented:
The framerate drop in the TRUB open area is fixed, thank you. 👍 

I've been checking the lightning in Folly/Atlantis and I noticed a couple of differences. The lightning is often hidden behind the altar in the PR build.

Left is develop, right is the PR build.

![Screenshot](https://github.com/user-attachments/assets/5c946ddb-a47c-460c-8261-9f8b9b8b6b54)

The lightning issue in Folly is difficult explain so I can make a video later, what I noticed is the lighting tend to always emit from the front side of the orb facing Lara, and not the back of it, whereas it would mostly emit from the back of the orb on develop.

Btw I was pleasantly surprised to discover that #2452 is fixed in this PR. 🎉 

____________

**Edit:** I noticed artifacts and a black contour on room sprites, but only when bilinear filter is enabled. The sprite is in City of Khamoon, room 22.

Left is develop, right is the PR build.

![Screenshot](https://github.com/user-attachments/assets/bf9e830b-3159-4fea-b7f0-58be21d21803)

rr- committed:
(wip) use separate vbo for the sprite uvs

rr- committed:
(wip) switch to cuboids; support animated textures

aredfan commented:
- [ ] ~~The `Save Game`/`Load Game` headers in the save/load dialogs are missing.~~ (present in develop)
- [ ] The room sprites are flickering, but only if bilinear filter is enabled.
- [ ] Parts of Lara doesn't show reflection when she stands on Midas' hand, if the reflection option is enabled.
- [ ] There's a subtle difference with the lighting. I wouldn't have noticed it until I started comparing screenshots between develop and the PR. I thought I'd mention it in case it's relevant.

Temple of the Cat, room 69

1st screenshot is the PR build (with the fps counter), 2nd screenshot is develop. Best to compare in fullscreen.

![20250316_213118_Temple_of_the_Cat](https://github.com/user-attachments/assets/1061074f-545b-4038-ad59-d1c43bb32543)
![20250316_213142_Temple_of_the_Cat](https://github.com/user-attachments/assets/87864656-39b2-4b97-83dd-cfd4b93970bb)

Richard-L commented:
@aredfan 
> Btw I was pleasantly surprised to discover that https://github.com/LostArtefacts/TRX/issues/2452 is fixed in this PR. 🎉

The issue is still present for me (UB cats) on [TR1X-4.8.3-208-g6e4f362-Windows.zip](https://github.com/LostArtefacts/TRX/releases/download/latest/TR1X-4.8.3-208-g6e4f362-Windows.zip)

rr- committed:
revert 4.3c trial

rr- commented:
Closing in favor of more incremental approach :relieved:
